Fix traceback forwarding with inlineCallbacks on python 3

tardyp opened the pull request:
fix for https://twistedmatrix.com/trac/ticket/9175
and https://twistedmatrix.com/trac/ticket/5949

tardyp committed:
fix inlineCallback traceback forwarding on py3

on PY3, b'a'[0] == 97 while in py2 b'a'[0] == b'a' opcodes are stored in bytes
this means that _findFailure was not able to find the yield opcode on py3

tardyp committed:
test for 9175

tardyp committed:
newsfragment

rodrigc reviewed (changes requested):
@tardyp instead of using _PY3, would it be possible to use `twisted.python.compat._bytesChr` here?

tardyp commented:
@rodrigc bytesChr will not work here as on python3 we need to have an int, and on py2 we need a single byte native string. bytesChr always return bytes https://github.com/twisted/twisted/blob/twisted-17.5.0/src/twisted/python/compat.py#L759

rodrigc commented:
OK.
There is a test failure with this patch: https://travis-ci.org/twisted/twisted/jobs/242081518

```
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/home/travis/build/twisted/twisted/build/py33-alldeps-withcov-posix/lib/python3.3/site-packages/twisted/test/test_failure.py", line 914, in test_throwExceptionIntoGenerator
    self.assertEqual(traceback.extract_tb(stuff[0][2])[-1][-1], "1/0")
  File "/home/travis/build/twisted/twisted/build/py33-alldeps-withcov-posix/lib/python3.3/site-packages/twisted/trial/_synctest.py", line 432, in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
  File "/opt/python/3.3.5/lib/python3.3/unittest/case.py", line 641, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/python/3.3.5/lib/python3.3/unittest/case.py", line 1020, in assertMultiLineEqual
    self.fail(self._formatMessage(msg, standardMsg))
twisted.trial.unittest.FailTest: 'yield' != '1/0'
- yield
+ 1/0
twisted.test.test_failure.ExtendedGeneratorTests.test_throwExceptionIntoGenerator
-------------------------------------------------------------------------------
```

jafd committed:
Pass the actual traceback to g.throw()

A quick check confirmed that http://twistedmatrix.com/trac/ticket/9175#comment:1 is
wrong that self.tb at the moment of throwing an exception back to the generator
is always None. Adding it back makes the failing tests pass.

jafd committed:
The tests for #5949 pass again, so no need to skip them anymore.

tardyp commented:
added test fixes from @jafd thanks to him. Sorry for forgetting about this PR :-/

rodrigc commented on src/twisted/test/test_failure.py:
```diff
@@ -933,14 +933,6 @@ def generator():
         self.assertEqual(len(newFailures), 1)
```
Please remove import of _PY3 from this file.  It is now an unused import

rodrigc commented on src/twisted/python/failure.py:
```diff
@@ -21,6 +21,7 @@
 from inspect import getmro
 
 from twisted.python.compat import NativeStringIO as StringIO
+from twisted.python.compat import _PY3
```
Combine this import with previous line

rodrigc commented on src/twisted/newsfragments/9175.bugfix:
```diff
@@ -0,0 +1 @@
+Fix traceback forwarding with inlineCallbacks on python 3.
```
Please add a newsfragment for 5949.  It looks like you fixed that as well.

rodrigc committed:
Fix imports

rodrigc commented:
@itamarst @exarkun @moshez @glyph Please review this

rodrigc committed:
Fix some whitespace and line length issues discovered by twistedchecker.

rodrigc commented on src/twisted/python/failure.py:
```diff
@@ -165,10 +171,17 @@ class Failure(BaseException):
+    if _PY3:
+        _yieldOpcode = opcode.opmap["YIELD_VALUE"]
+    else:
+        _yieldOpcode = chr(opcode.opmap["YIELD_VALUE"])
```
I've tested this on Python 2 and Python 3.  This is the right fix, because further down,
we dereference a single element in a bytes array

rodrigc commented on src/twisted/python/failure.py:
```diff
@@ -410,32 +430,33 @@ def _findFailure(cls):
+        # with no co_code, but they can't yield so we know it's okay to
+        # just return here.
         if ((not lastFrame.f_code.co_code) or
             lastFrame.f_code.co_code[lastTb.tb_lasti] != cls._yieldOpcode):
```
`lastFrame.f_code.co_code is an array of `bytes`.  On Python 2, `lastFrame.f_code.co_code[lastTb.tb_lasti]` returns a character, such as 'c'.  On Python 3, lastFrame.f_code.co_code[lastTb.tb_lasti]` returns an int, such as 99.
